README: Update the local building section (#15)
hdl/conda-compilers: README.md

## Testing conda builds locally

> Given a fairly empty *disposable* Ubuntu x86-64 test environment (e.g.,

Reviewer: It is recommend to build these packages in a fresh disposable environment such as a clean docker container.

While the goal is for the conda environment to be totally self contained, there is a constant battle to make sure this happens.

created with Docker or Vagrant), it is possible to simulate *part* of
what the CI does to test building individual toolchain architectures
locally.

The commands from the following subsections are enough to build a package
in a "clean" docker container based on the *ubuntu* Docker image.

### Prerequisites

The only required prerequisites are:
* [Git](https://git-scm.com/),
* Conda installed and initialized, e.g., using
[Miniconda](https://docs.conda.io/en/latest/miniconda.html)

> (it automatically makes required *python3* with *pip* and *setuptools*

Reviewer: ```suggestion
(which includes self contained versions of the required *python3* with *pip* and *setuptools* tooling)
```

available),
* [conda-build-prepare](https://github.com/litex-hub/conda-build-prepare),

> * this repository cloned to a local directory.

Reviewer: ```suggestion
this repository cloned to a local directory.
```

Author: I'm not sure you really meant because simply placing it after the list didn't look great. I started it with `and of course this repository...` to make it more natural.

All of these requirements can be satisfied using the following commands:

```bash
# Install git and wget (might require using `sudo`)
apt-get update
apt-get install -y git wget

# Download Miniconda and install in CONDA_PATH
CONDA_PATH=~/conda
wget -c https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
chmod a+x Miniconda3-latest-Linux-x86_64.sh
./Miniconda3-latest-Linux-x86_64.sh -p $CONDA_PATH -b -f

# Initialize Conda in the shell
eval "$("$CONDA_PATH/bin/conda" "shell.bash" 'hook' 2> /dev/null)"

# Install `conda-build-prepare`

> python3 -m pip install git+https://github.com/litex-hub/conda-build-prepare@v0.1#egg=conda-build-prepare

Reviewer: I wonder if this should be `pip install -e .` to use the locally cloned conda-build-prepare?

Author: I think it would only make sense if the `conda-build-prepare` will be modified and that's not the case for the most of the people building packages locally.

I guess a person who wants to modify a Python module knows or can easily find how to install it in the "editable" mode.

Reviewer 2: I agree with @ajelinski. Using `-e` is misleading for less experienced users, and the target developers/collaborators are likely aware of it. It might be added as a comment, but IMHO it should not be the default.

# Clone the litex-conda-compilers repository
git clone https://github.com/litex-hub/litex-conda-compilers.git
cd litex-conda-compilers
```

### Required environment variables

> Most importantly, the `RECIPE` variable must be set with the appropriate

Reviewer: This isn't quite true. The `RECIPE` argument is given to conda-build-prepare command as an argument?

Author: You're definitely right. I've redesigned this subsection to separate variables used in the script from information about the CBP arguments.

recipe's path to build the chosen package.

The recipe might require exporting some additional environment variables.
Such variables must be set before preparing the recipe, which is done
with one of the commands from the next subsection.

Currently required additional environment variables are:
* `TOOLCHAIN_ARCH` (by: `binutils`, `gcc/*`, `gdb` and `toolchain/linux-musl`
recipes) – must contain a supported architecture name, ie. one of these:
  * `lm32` (excluding `gcc/linux-musl` and `toolchain/linux-musl` recipes),
  * `or1k`,
  * `ppc64le` (only `binutils` and `gcc/linux-musl` recipes),
  * `riscv32`,
  * `riscv64`,
  * `sh`.

> Almost all of the recipes require exporting `DATE_NUM` and `DATE_STR`

Reviewer: `conda-build-prepare` should detect or produce reasonable default values for these so a user doesn't need to set them when building locally.

variables, since those are used as a build number and a build string

> respectively for the created package.

Reviewer: ```suggestion
respectively for the created package.

```

In the CI these variables are set based on the commit that triggered the
workflow (i.e., this repository's commit, not the recipe's git source).

```bash
# Set required variables with some values if they haven't been set already
RECIPE=${RECIPE:-binutils}

> if [ "$RECIPE" = "binutils"         \

Reviewer: I don't like this if statement here.

## Testing conda builds locally

It is recommend to build these packages in a fresh disposable environment

> such as a clean docker container.

Reviewer 2: I suggest: "*such as a clean container (Docker, Podman...)*".

While the goal is for the conda environment to be totally self contained,
there is a constant battle to make sure this happens.

The commands from the following subsections are enough to build a package

> in a clean docker container based on the *ubuntu* Docker image.

Reviewer 2: Is "any" version of Ubuntu supported? Or is is bionic only, for now?

I suggest using "*container image*" instead of "*Docker image*".

Author: Changed to "The commands from the following subsections were tested to be enough to build a package in a clean container based on `ubuntu` (`16.04`-`20.04`) or `debian` (`8`-`10`) Docker image." after I tested them on these OSs.

### Prerequisites

The only required prerequisites are:
* [Git](https://git-scm.com/),
* Conda installed and initialized, e.g., using
[Miniconda](https://docs.conda.io/en/latest/miniconda.html)
(which includes self contained versions of the required *python3* with
*pip* and *setuptools* tooling),
* [conda-build-prepare](https://github.com/litex-hub/conda-build-prepare)

> and of course this repository cloned to a local directory.

Reviewer 2: I would try avoiding "*of course*" and such. I also use them more than I should, but they don't help much.

I suggest:

"*Apart from cloning this repository to a local directory, the following prerequisites are required:*"

Author: You're right, it's much better.

> All of these requirements can be satisfied using the following commands:

Reviewer 2: "*On Debian/Ubuntu, these requirements can be satisfied...*"

```bash
# Install git and wget (might require using `sudo`)
apt-get update
apt-get install -y git wget

# Download Miniconda and install in CONDA_PATH
CONDA_PATH=~/conda

> wget -c https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh

Reviewer 2: I would write L120-L122 in a single line:

```sh
curl -fsSL https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh | sh -s -- -p $CONDA_PATH -b -f 
```

Author: Generally it's a nice idea but unfortunately this script prohibits sourcing it. It can only be executed so this doesn't work.

chmod a+x Miniconda3-latest-Linux-x86_64.sh
./Miniconda3-latest-Linux-x86_64.sh -p $CONDA_PATH -b -f

# Initialize Conda in the shell
eval "$("$CONDA_PATH/bin/conda" "shell.bash" 'hook' 2> /dev/null)"

# Install `conda-build-prepare`
python3 -m pip install git+https://github.com/litex-hub/conda-build-prepare@v0.1#egg=conda-build-prepare

# Clone the litex-conda-compilers repository
git clone https://github.com/litex-hub/litex-conda-compilers.git
cd litex-conda-compilers
```

### Required environment variables

> The recipe might require exporting some additional environment variables.

Reviewer 2: Which recipe? "The Conda recipe"? Which one?

[These are socratic questions]

Author: I've changed it to "Some recipes require exporting additional environment variables."

Such variables must be set before preparing the recipe, which is done
with one of the commands from the next subsection.

Currently required additional environment variables are:

> * `TOOLCHAIN_ARCH` (by: `binutils`, `gcc/*`, `gdb` and `toolchain/linux-musl`

Reviewer 2: Maybe not as part of this PR, but this feels like a good candidate for a `configure` shell script.

Author: `TOOLCHAIN_ARCH` isn't about the host architecture so I'm not sure what would be the purpose of such a script in this case.

recipes) – must contain a supported architecture name, ie. one of these:
  * `lm32` (excluding `gcc/linux-musl` and `toolchain/linux-musl` recipes),
  * `or1k`,
  * `ppc64le` (only `binutils` and `gcc/linux-musl` recipes),
  * `riscv32`,
  * `riscv64`,
  * `sh`.

The optional `DATE_NUM` and `DATE_STR` environment variables are used by the
recipes to set the `build/number` and `build/string` keys during preparation.
In case they're unset, recipe preparation provides them with default values

> based on the latest commit's committer date from the repository holding the

Reviewer 2: Some users might attempt to download an archive of the repos, instead of cloning them through git. If having the git history is required (which I find sensible), I believe it'd be interesting to specify it above.

Author: You're right. I didn't think enough about downloading the repository as an archive. I stressed such a necessity in the description and added the commented `export DATE_*` commands to the script.

I will consider providing them automatically even if there's no git repository, as this can be more common than I thought initially.

recipe.

### Preparing and building the package

> After getting all prerequisites and setting the required variables, it

Reviewer 2: I feel there might be misunderstandings about what "local" means. There are three possible setups:

- The "native" Ubuntu environments.
- A container.
- Executing some commands natively and others in a container.

Those three setups are possible either locally or in CI. Hence, 6 total. At first, I thought that you were only considering "native Ubuntu environments" in CI and "a container" locally. Then, I understood that repos might be cloned natively, and other commands executed in a container. Now, I am unsure.

My recommendation would be to create/maintain an specific container image with all the prerequisites. Then, use that image both in CI and locally. That will reduce startup time, and it will provide exactly the same result for all developers/collaborators and CI services.

It might make sense to create litex-hub/containers or symbiflow/containers for maintaining build images. We own the symbiflow namespace in dockerhub (I can transfer/give access to any of you). See ghdl/docker (and ghdl/ghdl) and hdl/containers for practical examples.

Author: All of the commands provided have to be executed in the same OS and are the same regardless whether used in a host OS or a container. The commands have been tested mostly on Debian/Ubuntu, but apart from the package installation tool there shouldn't be many problems to use it on many other GNU/Linux OSs.

Did command separation into two blocks made you think that some commands should be executed in a host OS and the others in a container? In the same sentence where you suggested a change there is: 
> The commands from the following subsections are enough to build a package in a clean docker container based on the *ubuntu* Docker image.

but maybe I should stress that the commands should also be executed there? Maybe you want to suggest some change to make everything clear?

Your recommendation seems to be very reasonable but this would be quite a change. I don't think it should be handled by this PR. Maybe you could create an issue pointing out what are the problems with the current situation and your suggestion? Simply to not forget about it as I think it is a good idea.

is recommended to prepare the recipe before building, as it gives you
the advantages described on [the `conda-build-prepare`'s
GitHub page](https://github.com/litex-hub/conda-build-prepare).
The tool is also used within the CI, which makes the locally built
package much more similar to the one built by the CI workflow.

> The tool is a Python module.

Reviewer 2: I would suggest being more explicit than "*the tool*". Maybe "*the entrypoint to this build tooling is a Python module with a CLI*".

Author: You're right but I changed it to "The `conda-build-prepare` is a Python module with a CLI."

Author: You might've been confused about which tool this was as there was "the tool". I made the whole paragraph much more precise now.

> The only required arguments are the path to the recipe corresponding to the

Reviewer 2: Maybe out of scope of this PR, but it'd be interesting to generate this details as a canonical help (docstrings) in Python. Then, copy the output of `-h` here.

Paebbels/pyAttributes might be useful.

Author: I don't think that users should really be informed about all the details of `conda-build-prepare`. If the primary goal is to make building process similar to the CI one then it's best when the same arguments as in that CI are used.

Didn't know the pyAttributes though, I'll check it out.

Author: You might've been confused about what tool this was as there was "the tool". I made the whole paragraph much more precise now.

Reviewer 2: My point is, are users expected to execute `conda-build-prepare` or is it for developers only? If it's going to be executed by users, I think that a list output is desirable:

- `--dir`: ...
- `--channels`: ...
- `--packages`: ...

By using pyAttributes or a similar solution, the amount of info shown to users can be decided.

Author: Anybody who wants to build a package as it's done in the CI should use *conda-build-prepare* and with the `--channels` and `--packages` the same as in README (=CI).

I've changed the whole section on *conda-build-prepare* arguments and I think it's much more comprehensible now. I've explained thoroughly the options that user might be interested in, what should be passed to the remaining ones and why.

# export DATE_NUM="$(date -u +%Y%m%d%H%M%S)"
# export DATE_STR="$(date -u +%Y%m%d_%H%M%S)"
#
# Or lock to date/time of the last commit on git, as it's done in the CI:
#
DATE_TS="$(git log --format=%ct -n1)"
export DATE_NUM="$(date --date=@${DATE_TS} -u +%Y%m%d%H%M%S)"
export DATE_STR="$(date --date=@${DATE_TS} -u +%Y%m%d_%H%M%S)"
```

### Preparing and building the package

After getting all prerequisites and setting the required variables, it
is recommended to prepare the recipe before building, as it gives you
the advantages described on [the `conda-build-prepare`'s
GitHub page](https://github.com/litex-hub/conda-build-prepare).
The tool is also used within the CI, which makes the locally built
package much more similar to the one built by the CI workflow.

> The `WORKDIR` variable is the name of the directory that will be

Reviewer: I don't like the name `WORKDIR` -- I would prefer something more descriptive like `FINALIZE_PACKAGE_OUTPUTDIR` or something along those lines?

Author: I've changed it to `PREPARED_RECIPE_OUTPUTDIR` since it's a path of the new directory that will contain output after preparing the recipe. Of course I'm open for modifications if you think it should stress the fact that it's finalized or something else. I think no name is fully accurate apart from the longish and less descriptive `CONDA_BUILD_PREPARE_OUTPUTDIR. Perhaps `CBP_OUTPUTDIR` could be used but it's still less descriptive.

created  as an output of running `conda-build-prepare`.

> It will contain these subdirectories:

Reviewer: Can you explain what is in each of these directories?

* `conda-env`,
* `git-repos` (if the recipe contains any git sources),
* `recipe`.
`WORKDIR` can thus be any valid directory name.

The arguments passed with `--channels` and `--packages` switches to the
`conda-build-prepare` are similar to those used within the CI workflow.

```bash
# Prepare the RECIPE with `conda-build-prepare`

> ADDITIONAL_PACKAGES="conda-build=3.20.3 conda-verify jinja2 pexpect python=3.7"

Reviewer: I'm confused about why `ADDITIONAL_PACKAGES` is needed?

Author: This is what we use in the CI so if we seek the maximum similarity with the CI then locally these should also be used.
I'm quite sure that only these set with versions make any difference so the rest (`conda-verify jinja2 pexpect`) can be dumped both from the CI and from here. Should I do it?

Author: In the end I've removed `conda-verify jinja2 pexpect` from here as I'm pretty sure they have zero influence on building.

`conda-build` and `python` have been left with versions for maximal CI resemblence.
